lib/pbench/cli/server/shell.py

```diff
@@ -37,6 +42,72 @@ def find_the_unicorn(logger: Logger):
             )
 
 
+def keycloak_connection(config: PbenchServerConfig, logger: Logger) -> int:
+    """
+     Checks if the Keycloak server is up and accepting the connections.
+     The connection check does the GET request on the oidc server /health
+     endpoint and the sample response returned by the /health endpoint looks
+     like the following:
+        {
+            "status": "UP", # if the server is up
+            "checks": []
+        }
+     Note: The Keycloak server needs to be started with health-enabled on.
+     Args:
+        config: PbenchServerConfig
+        logger: logger
+    Returns:
+        0 if successful
+        1 if unsuccessful
+    """
+    try:
+        oidc_server = config.get(
+            "authentication",
+            "internal_server_url",
+            fallback=config.get("authentication", "server_url"),
+        )
+    except (NoSectionError):
+        logger.exception("Bad config file:  missing 'authentication' section")
+        return 1
+    except (NoOptionError):
+        logger.error("Bad config file:  no 'server_url' in  'authentication' section")
+        return 1
+
+    session = requests.Session()
+    # The connection check will retry multiple times unless successful, viz.,
+    # [0.0s, 4.0s, 8.0s, 16.0s, ..., 120.0s]. urllib3 will sleep for:
+    # {backoff factor} * (2 ^ ({number of total retries} - 1)) seconds between
+    # the retries. However, the sleep will never be longer than backoff_max
+    # which defaults to 120.0s More detailed documentation on Retry and
+    # backoff_factor can be found at:
+    # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
+    retry = Retry(
+        total=8,
+        backoff_factor=2,
+        status_forcelist=tuple(int(x) for x in HTTPStatus if x != 200),
+    )
+    adapter = HTTPAdapter(max_retries=retry)
+    session.mount("http://", adapter)
+
+    # We will also need to retry the connection if the health status is not UP
+    for _ in range(5):
+        try:
+            response = session.get(f"{oidc_server}/health")
+            response.raise_for_status()
+        except Exception:
+            logger.exception("Error connecting to the OIDC client")
+            return 1
+        if response.json()["status"] == "UP":
```

Comment:
We can but it will always return the `status` field in the JSON payload if the return status code is 200 and we check the `status` only if the return code is 200. https://github.com/keycloak/keycloak/blob/main/docs/guides/src/main/server/health.adoc

Reply:
> it will always return the status field in the JSON

You mean "if it works as documented, then it will...".  🙂 

The question that Peter is (implicitly) posing is, what do we want the Pbench Server's behavior to be if Keycloak fails to honor its interface contract?  As coded, the Server startup will die with a traceback from a mysterious `KeyError` exception for `"status"`.  Is that acceptable (especially given the likelihood of this ever happening), or should we implement steps to report this situation in a way which would be easier to understand.

I'm thinking that, given the likelihood of failure and given the fact that it will fail long before the Server starts, we can ignore this and go with the code as is.

Reply:
I updated it from dict `[ ]` to `get` so that should take care of it now.

Reply:
Yeah, that seems like a reasonable compromise.